[DNF5] Consolidate DNF vars into a Vars class

lukash opened the pull request.

Commit:
Introduce the Vars class and consolidate the substitutions functionality

Commit:
test/libdnf/utils: Move to_vector(...) definition to a .cpp file

Commit:
Add tests for the Vars class

Commit:
Add basic tests for the conf module

Commit:
Remove ConfigRepo from Repo constructor

No need to pass it, the Repo can create it itself, pulling ConfigMain
from base it already got passed. Passing the ConfigRepo as a unique_ptr
rvalue is not a nice interface either.

Commit:
Add const getters to Config classes

These allow to pass around const references to configs for read-only
access to their options.

Commit:
Don't wrap Repo::Impl::conf in unique_ptr and rename to "config"

Changes the Repo::Impl::conf member to be a regular variable instead of
being wrapped in a unique_ptr, which is unnecessary. Also renames it
from "conf" to "config" for consistency with Base::config.

(All the changes are bascially replacing "conf->" with "config.",
no unrelated changes in the big diff)

Commit:
tests/libdnf/rpm/test_reldep_list: Use RepoFixture

Commit:
repo: Return reference instead of pointer from get_config()

This unifies the interface with Base::get_config(), returning a
reference to ConfigRepo instead of a pointer.

lukash commented:
I've added the commits for adding the const variants for Config getters and some interface cleanup (it's a bit big, but simple changes for the most part).

jrohel commented on include/libdnf/conf/config.hpp:
```diff
@@ -20,15 +20,49 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 public:
     OptionBinds & opt_binds() noexcept { return binds; }
 
+    void load_from_parser(
```
1. Could you move the implementation of the method to the .cpp file?
    And move `#include <fmt/format.h>` to .cpp file. (btw It solves problem with compiling swig ruby bindings on my PC)

lukash replied:
Done.

jrohel commented on include/libdnf/conf/config.hpp:
```diff
@@ -20,15 +20,49 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 namespace libdnf {
 
 /// Base class for configurations objects
+template <Option::Priority default_priority>
```
Does it have to be a template? What about an internal variable, adjustable e.g. in the constructor. Yes, each instance will be larger by this variable. But it saves a piece of code. And again there will be one class `Config` from which the others will inherit. Instead of multiple parent classes generated from a template.

Commit:
Fix ReldepList::add_reldep_with_glob()

Filter the searched libsolv data for reldep attributes only (without
this, the Dataiterator returns any string attributes that match the
input).

lukash commented:
@jrohel the PR is cleaned up, I've added a comit that fixes the reldep list issue too.

lukash commented on libdnf/rpm/reldep_list.cpp:
```diff
@@ -95,8 +95,23 @@ bool ReldepList::add_reldep_with_glob(const std::string & reldep_str) {
+                    dep_splitter.get_evr_cstr(),
+                    dep_splitter.get_cmp_type()
+                ));
+        }
```
@j-mracek here's my fix of listing the libsolv attributes, in case you want to have a look. I think I got it right.

j-mracek commented on libdnf/rpm/reldep_list.cpp:
```diff
@@ -95,8 +95,23 @@ bool ReldepList::add_reldep_with_glob(const std::string & reldep_str) {
-        ReldepId id = Reldep::get_reldep_id(sack, di.kv.str, dep_splitter.get_evr_cstr(), dep_splitter.get_cmp_type());
-        add(id);
+        switch (di.key->name) {
+            case SOLVABLE_PROVIDES:
```
And what about SOLVABLE_NAME?

lukash replied:
Is it not added to provides automatically?

j-mracek replied:
No. Package libdnf.s390 do not provide anything on x86_64 system. Same for src packages.
The other question would be, is it a problem? We suggest that such a reldep will be used for queries therefore anyway you will be not able to match anything.

lukash replied:
Right. I don't know, the way I understand it, this method is quite special. It depends on the intended semantics, its docstring is not very clear, it should say something like:
> Matches `reldep_str` against all reldep data in the pool and adds all matching reldeps to the list.

The method is on the interface, even though it's internally used in queries doesn't mean someone can't find an inventive use outside that.

I think the special nature of the method warrants it rather shouldn't be an "add" method of the ReldepList, I think it should be a method on perhaps the Sack (maybe elsewhere?) like this:
```
ReldepList Sack::get_all_reldeps_matching_glob(const std::string & glob);
```

The reasoning is the ReldepList should primarily be a container with simple methods to manipulate the contents, clever logic like the one being done here should be somewhere up the class hierarchy.

That said, I'm still not sure about the name. I guess it is a reldep identifier and so should be part of the list?

j-mracek replied:
I suggest that we can do with the present implementation.

lukash replied:
We've agreed to leave out SOLVABLE_NAME for now, so this should be good as it is.

jrohel commented:
@lukash Please do not mix unrelated things into this PR. The name of this PR is "Consolidate DNF vars into a Vars class". Please move the commit "Fix ReldepList :: add_reldep_with_glob ()" to the new PR. And if a test goes wrong, ban it and write TODO there. You will fix it in the new PR. Together with "Fix ReldepList :: add_reldep_with_glob ()" if related.

j-mracek commented on microdnf/context.cpp:
```diff
@@ -282,10 +282,10 @@ void Context::load_rpm_repos(libdnf::rpm::RepoQuery & repos, libdnf::rpm::SolvSa
 //             load_rpm_repo(*repo.get());
 //             auto & solv_sack = base.get_rpm_solv_sack();
-//             // std::cout << "Loading repository \"" << repo->get_config()->name().get_value() << "\" into sack." << std::endl;
+//             // std::cout << "Loading repository \"" << repo->get_config().name().get_value() << "\" into sack." << std::endl;
```
Please could you add todo why the code was left here?

lukash replied:
What TODO? I didn't comment out the code, I just updated it even though it's commented out (in an attempt to keep it working).

If you look at the first line of the comment, this is the single-threaded version that @jrohel left here, I guess mainly it htere were problems with the multi-threaded one, for debugging purposes... it's not ideal, as it will rot over time, maybe we should consider keeping the single-threaded version as an active alternative conde-path (using a compile-time or a run-time switch... maybe an env variable?)

j-mracek replied:
I know, but please could you add it to correct it. When we left some code commented out we should mark it with todo to not forget about. Thanks a lot.

lukash replied:
Allright, I guess I can add something.

lukash commented:
> @lukash Please do not mix unrelated things into this PR. The name of this PR is "Consolidate DNF vars into a Vars class". Please move the commit "Fix ReldepList :: add_reldep_with_glob ()" to the new PR. And if a test goes wrong, ban it and write TODO there. You will fix it in the new PR. Together with "Fix ReldepList :: add_reldep_with_glob ()" if related.

I can't move the commits to a new PR until this one gets merged (as that PR would just contain all the commits in this one). I've had the code ready, so I added it to this PR to make things easier and faster... Unless you want to merge only a part of this PR, it is now easier to keep it the way it is?

(I can rename it, as the name is no longer accurate...)

Commit:
microdnf: Add a TODO to single-threaded repo loading implementation

lukash commented:
@jrohel rebased and the TODO @j-mracek wanted is added, I think it's ready to merge.

jrohel commented:
OK. Conversation with j-mracek is solved too. I will merge it.
